source/fuzz/shrinker.cpp

```diff
@@ -0,0 +1,223 @@
+    impl_->consumer(SPV_MSG_INFO, nullptr, {},
+                    "Initial binary is invalid; stopping.");
+    return Shrinker::ShrinkerResultStatus::kInitialBinaryInvalid;
+  }
+
+  std::vector<uint32_t> current_best_binary;
+  protobufs::TransformationSequence current_best_transformations;
+
+  // Run a replay of the initial transformation sequence to (a) check that it
+  // succeeds, (b) get the binary that results from running these
+  // transformations, and (c) get the subsequence of the initial transformations
+  // that actually apply (in principle this could be a strict subsequence).
+  if (Replayer(impl_->target_env)
+          .Run(binary_in, initial_facts, transformation_sequence_in,
+               &current_best_binary, &current_best_transformations) !=
+      Replayer::ReplayerResultStatus::kComplete) {
+    return ShrinkerResultStatus::kReplayFailed;
+  }
+
+  // Check that the binary produced by applying the initial transformations is
+  // indeed interesting.
+  if (!interestingness_function(current_best_binary, 0)) {
+    impl_->consumer(SPV_MSG_INFO, nullptr, {},
+                    "Initial binary is not interesting; stopping.");
+    return ShrinkerResultStatus::kInitialBinaryNotInteresting;
+  }
+
+  uint32_t attempt = 0;  // Keeps track of the number of shrink attempts that
+                         // have been tried, whether successful or not.
+
+  uint32_t granularity =
+      std::max(1u, NumRemainingTransformations(current_best_transformations) /
+                       2);  // The number of contiguous transformations that the
+                            // shrinker will try to remove in one go; starts
+                            // high and decreases during the shrinking process.
+
+  // Keep shrinking until we:
+  // - reach the step limit,
+  // - run out of transformations to remove, or
+  // - cannot make the granularity of shrinking any finer.
+  while (attempt < impl_->step_limit &&
+         !current_best_transformations.transformation().empty() &&
+         granularity > 0) {
+    bool progress_this_round =
+        false;  // Used to decide whether to make the granularity with which we
+                // remove transformations finer.  If we managed to remove at
+                // least one chunk of transformations at a particular
+                // granularity, we set this flag so that we do not yet decrease
+                // granularity.
+
+    // We go through the transformations in reverse, in chunks of size
+    // |granularity|. This is encoded here via an index, |index|, recording how
+    // many chunks of size |granularity| have been tried. The loop exits early
+    // if we reach the shrinking step limit.
+    for (uint32_t index = 0;
+         attempt < impl_->step_limit &&
+         index * granularity <
+             NumRemainingTransformations(current_best_transformations);) {
+      // Remove a chunk of transformations according to the current index and
```

Comment:
In other versions of the reducer algorithm: we had a check at the start of the loop that reduces `granularity` if it is more than the number of transformations/opportunities. Here, we could do:
if granularity >= len(transformations): granularity = max(1, len(transformations) / 2)

Reply:
We also usually reset the index in this case.

Reply:
I have reworked the shrinker algorithm quite a bit in this revision.  I believe it is somewhat simpler now, and it's now more evident that if we remove chunks in reverse order, removing a later chunk cannot affect whether an earlier chunk still applies, so long as the points at which chunks are deemed to start is not affected.